### DISCUSSION TITLE
Cookies not being generated except on localhost

### DISCUSSION THREAD

**Discussion by itrase** [danny-avila/LibreChat#471] [Troubleshooting] [answered]
I'm very new to Javascript/web development, so this may be an obvious issue. I've set up LibreChat and can access it both from a localhost address and from the public IPv4 address. When I login using the localhost address, a cookie is created and saved to the browser, such that when I refresh the page, I am still logged in. However, when I log in using the public IPv4 address, a cookie is not created, and so if I refresh the page, I am forced to log in again. Is there some setting I am missing in one of the two .env files or elsewhere?

**Comment by danny-avila** [comment]
@itrase could be a couple things in your local environment.

how are you accessing the address in your browser? are you using `https://` or `http://` ?

**Reply by itrase** [reply]
I think I'm using http://. When I put "http://XX.XX.XX.XX:3080/new/chat" into the browser, it redirects to "XX.XX.XX.XX:3080/login"

**Reply by danny-avila** [reply]
are you using google login? you would need to add your public ip to the ~~redirect~~ Authorized JavaScript origins in your google cloud settings, and make sure the env variable is set to the correct Environment (production will point to your prod url, dev to dev url)

Try 'https' instead of 'http' (your deployment may need to handle https somehow and your browser may be blocking http cookies for public addresses)

Using ngrok, and the right settings, i have no problem with cookies

**Reply by danny-avila** [reply]
and login state is persisted

**Reply by danny-avila** [reply]
Last thing, you would need to log in first for each endpoint. Your localhost cookies won't persist into the public IP address

**Reply by itrase** [reply]
We aren't using google login - this is a small internal test, so we don't really have the ability to stand up https. I've just been running the app by running "docker-compose up" on an EC2 instance and connecting from there (so no ngrok or anything else). As far as I know, chrome is not blocking 3rd party cookies - it's set to only block cookies in Incognito.

**Reply by danny-avila** [reply, marked as answer]
what is your node env in the ec2 instance? The cookies are set to secure in a production environment and this is probably what's happening. See this for more info: https://stackoverflow.com/questions/73752871/aws-ec2-instance-does-not-set-cookie-when-sent-by-node-server-running-inside-it

You said you're using docker. the default NODE_ENV on the compose file is 'production'

edit this line in your docker-compose.yml file `      - NODE_ENV=production` to `      - NODE_ENV=development`

**Reply by itrase** [reply]
Yes, setting the NODE_ENV to development in the docker-compose.yml worked - thank you so much!!

**Reply by danny-avila** [reply]
Awesome! Thanks for reaching out and for trying out the project! Hope it serves your purposes

**Reply by mycomp32** [reply]
i want to leave it in production and enable ssl, is that the way to have cookies working? if so, where can i enable ssl, i have the certs

**Reply by danny-avila** [reply]
I haven't done this but my guess would be you need to edit api/server/index.js 

https://stackoverflow.com/questions/11744975/enabling-https-on-express-js

**Reply by sisland** [reply]
Hi all!

I am using docker too on a different machine in my home, but when using from other than localhost (lets say something like "HTTP://192.168.0.250" ), cookies are not being saved. I tried your solution on my docker-compose.yml inside the LibreChat folder:

`
services:
  api:
    container_name: LibreChat
    ports:
      - "${PORT}:${PORT}"
    depends_on:
      - mongodb
      - rag_api
    image: ghcr.io/danny-avila/librechat-dev:latest
    restart: always
    user: "${UID}:${GID}"
    extra_hosts:
      - "host.docker.internal:host-gateway"
    environment:
      - HOST=0.0.0.0
      - MONGO_URI=mongodb://mongodb:27017/LibreChat
      - MEILI_HOST=http://meilisearch:7700
      - RAG_PORT=${RAG_PORT:-8000}
      - RAG_API_URL=http://rag_api:${RAG_PORT:-8000}
      - NODE_ENV=development`
      
Those changes were done with docker stopped.  After that, I applied:
`sudo docker-compose -f ./deploy-compose.yml up -d`
But seems isn't working as expected.

I am doing something wrong? I am using latest image of Librechat on Docker 

Thanks!

**Reply by danny-avila** [reply]
@sisland please see the discussion here: https://github.com/danny-avila/LibreChat/discussions/572

**Reply by danny-avila** [reply]
TL;DR you need HTTPS or run in dev mode.

**Reply by jaylee2019** [reply]
> what is your node env in the ec2 instance? The cookies are set to secure in a production environment and this is probably what's happening. See this for more info: https://stackoverflow.com/questions/73752871/aws-ec2-instance-does-not-set-cookie-when-sent-by-node-server-running-inside-it
> 
> You said you're using docker. the default NODE_ENV on the compose file is 'production'
> 
> edit this line in your docker-compose.yml file ` - NODE_ENV=production` to ` - NODE_ENV=development`

I have the same issue. This method is not work for me. Is there any other solutions?

**Reply by TomasJayess** [reply]
> what is your node env in the ec2 instance? The cookies are set to secure in a production environment and this is probably what's happening. See this for more info: https://stackoverflow.com/questions/73752871/aws-ec2-instance-does-not-set-cookie-when-sent-by-node-server-running-inside-it
> 
> You said you're using docker. the default NODE_ENV on the compose file is 'production'
> 
> edit this line in your docker-compose.yml file ` - NODE_ENV=production` to ` - NODE_ENV=development`

I'm currently trying to work on this in 2025, but I can't find the line "NODE_ENV=x" in either the docker-compose.yml or docker-compose.override.yml. is there another solution to this? Am I just somehow missing the line? 

Edit:  I found a "NODE_ENV" field in the "deploy-compose.yml" and changed it from "production" to "development" I don't think it changed anything.

**Reply by OPerepadia** [reply]
> I found a "NODE_ENV" field in the "deploy-compose.yml" and changed it from "production" to "development" I don't think it changed anything.

I don't think `deploy-compose.yml` is the right file.
What worked for me is setting  `NODE_ENV=development` in `.env` file

**Reply by TomasJayess** [reply]
Yeah I found that after I posted this. Still didn't seem to make a difference, but I'm going back to undo the thing deploy-compose edit just in case.

**Comment by mycomp32** [comment]
got it working just used nginx ssl proxy to fwd to localhost

**Comment by dazipe** [comment]
The suggested fix doesn't work anymore.
I checked the docker compose logs like this:

`docker compose logs api`

 And noticed that by default the api container is started with hardcoded:
  NODE_ENV=production node api/server/index.js

So i had to modify the docker-compose.yaml
```
services:
  api:
    container_name: LibreChat
    ports:
      - "${PORT}:${PORT}"
    depends_on:
      - mongodb
      - rag_api
    image: ghcr.io/danny-avila/librechat-dev:latest
    restart: always
    user: "${UID}:${GID}"
    extra_hosts:
      - "host.docker.internal:host-gateway"
    command: sh -c "NODE_ENV=development node api/server/index.js"

```
the last line command: sh -c "NODE_ENV=development node api/server/index.js" did the trick.
